### PR TITLE
[JUJU-268] Add support for tear down of instance roles.

### PR DESCRIPTION
**Pull request by tlm**
With instance role credentials the controller cannot use the model credentials to assist in tear down. This introduces a change where if this is detected we swap to using the controller credentials. This is a best effort approach at the moment

## Checklist

 - [x] Requires a [pylibjuju](https://github.com/juju/python-libjuju) change
 - [x] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed
 - [x] Comments answer the question of why design decisions were made

## QA steps

Bootstrap to AWS using:

```sh
juju bootstrap --bootstrap-constraints="instance-role=auto" aws/ap-southeast-2 tlm-test-1
```

Then kill the newly created controller with:
```bash
juju kill-controller -t 0 -y tlm-test-1
```

## Documentation changes

N/A

## Bug reference

N/A

### CODE REVIEW

File: cmd/juju/controller/destroy.go [hunk 1]
```diff
@@ -590,6 +590,21 @@ func (c *destroyCommandBase) getControllerEnviron(
 	return env, nil
 }
 
+func (c *destroyCommandBase) getControllerCloudSpecFromStore(
+	ctx *cmd.Context,
+	store jujuclient.ClientStore,
+	controllerName string,
+) (environscloudspec.CloudSpec, error) {
+	_, params, err := modelcmd.NewGetBootstrapConfigParamsFunc(
```
Comment (wallyworld) [inline cmd/juju/controller/destroy.go:598]: This means that the command will only work when run from the exact same client that was used to bootstrap. And these client bootstrap params we are trying to get rid of. Can't we make an api call to the controller to get the cloud info off the controller model? This allows any juju client to run the kill command.

File: environs/interface.go [hunk 2]
```diff
@@ -288,6 +288,14 @@ type ConfigSetter interface {
...
 }
 
+// CloudSpecGetterFn provides a function type that implements CloudSpecGetter
+type CloudSpecGetterFn func(stdcontext.Context) (environscloudspec.CloudSpec, error)
+
+// GetCloudSpec implements CloudSpecGetter interface
+func (c CloudSpecGetterFn) GetCloudSpec(ctx stdcontext.Context) (environscloudspec.CloudSpec, error) {
+	return c(ctx)
+}
+
```
Comment (wallyworld) [inline environs/interface.go:298]: delete me

File: provider/ec2/environ.go [hunk 3]
```diff
@@ -56,6 +56,8 @@ const (
 	// AWSClientContextKey defines a way to change the aws client func within
 	// a context.
 	AWSClientContextKey corecontext.ContextKey = "aws-client-func"
+
+	AWSIAMClientContextKey corecontext.ContextKey = "aws-iam-client-func"
```
Comment (wallyworld) [inline provider/ec2/environ.go:60]: Comment?

**Review by wallyworld — APPROVED** [review]
As pointed out to me live, the controller cloud info stored in the model is altered from what was used to bootstrap. So we do need to use what's on the client. Let's add a nice error message if the necessary bootstrap params are missing. That plus the missing test and small cleanup and I think we are good.

**Comment by tlm** [comment]
$$merge$$

**Pull request merged**